assembly/src/tokens/stream.rs

```diff
@@ -22,63 +24,65 @@ impl<'a> TokenStream<'a> {
     pub fn new(source: &'a str) -> Result<Self, ParsingError> {
         // initialize the attributes
         let mut tokens = Vec::new();
-        let mut lines = Vec::new();
+        let mut locations = Vec::new();
         let mut proc_comments = BTreeMap::new();
         let mut module_comment = None;
 
-        // fetch all tokens
-        for line_info in LinesStream::from(source) {
-            let line_number = line_info.line_number() as usize;
-
-            match line_info.contents() {
-                Some(line) => {
-                    // fill the doc comments for procedures
-                    if line.starts_with(Token::EXPORT) || line.starts_with(Token::PROC) {
-                        let doc_comment = build_comment(line_info.docs());
-                        proc_comments.insert(tokens.len(), doc_comment);
-                    } else if !line_info.docs().is_empty() {
-                        return Err(ParsingError::dangling_procedure_comment(line_number));
-                    }
-
-                    // for each token, skip comments & err when dangling docs; push otherwise
-                    for token in line.split_whitespace() {
-                        if token.starts_with(Token::DOC_COMMENT_PREFIX) {
-                            return Err(ParsingError::dangling_procedure_comment(line_number));
-                        } else if token.starts_with(Token::COMMENT_PREFIX) {
-                            break;
-                        }
+        for info in LinesStream::from(source) {
+            let mut docs = info.docs();
+            let mut tokenizer = LineTokenizer::from(&info);
+
+            // first dangling docs are module docs
+            if tokens.is_empty() && info.contents().is_none() {
+                let docs = mem::take(&mut docs);
+                let docs = build_comment(docs);
+                module_comment = docs;
+            } else if info.contents().is_none() {
+                let location = SourceLocation::new(info.line_number(), 1);
+                return Err(ParsingError::dangling_procedure_comment(location));
+            }
```

Comment:
Could we not do something simpler here? Maybe something like this:
```Rust
for line_info in LinesStream::from(source) {
    match line_info.contents() {
        Some(line) => {
            // fill the doc comments for procedures
            if line.starts_with(Token::EXPORT) || line.starts_with(Token::PROC) {
                let doc_comment = build_comment(line_info.docs());
                proc_comments.insert(tokens.len(), doc_comment);
            } else if !line_info.docs().is_empty() {
                return Err(ParsingError::dangling_procedure_comment(line_info.into()));
            }

            // break the line into tokens and record their locations
            let mut tokenizer = LineTokenizer::from(&line_info);
            for (token, location) in tokenizer.by_ref() {
                tokens.push(token);
                locations.push(location);
            }

            // if the line ends with a procedure doc comment, return an error
            if let Some(location) = tokenizer.take_dangling() {
                return Err(ParsingError::dangling_procedure_comment(location));
            }
        }

        // if first dangling comment, then module docs
        None if tokens.is_empty() => {
            module_comment = build_comment(line_info.docs());
        }

        // if has tokens, then dangling docs are illegal
        None => {
            return Err(ParsingError::dangling_procedure_comment(line_info.into()));
        }
    }
}
```
Assuming it works, I think it is both easier to follow and has fewer changes to the current code.

Reply:
Updated with the suggested snippet here: https://github.com/0xPolygonMiden/miden-vm/pull/861/commits/46bfb7a92b1b502846f09aa8d9a987bb394d3db8